src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -929,7 +929,19 @@ public void executorInit(CommandEnvironment env, BuildRequest request, ExecutorB
       remoteOutputService.setActionInputFetcher(actionInputFetcher);
       actionContextProvider.setActionInputFetcher(actionInputFetcher);
 
-      if (remoteOutputsMode.downloadToplevelOutputsOnly()) {
+      boolean useToplevelDownloader = remoteOutputsMode.downloadToplevelOutputsOnly();
+
+      // Download toplevel artifacts for `run` command.
+      if ("run".equals(env.getCommandName())) {
+        useToplevelDownloader = true;
+      }
+
+      // Download toplevel artifacts for `coverage` command.
+      if ("coverage".equals(env.getCommandName())) {
```

Comment:
@coeuvre This looks great, thanks for picking it up! 

One thing I noticed: Could this make `bazel coverage` download not only the coverage reports, but also the test binaries itself (see https://github.com/bazelbuild/bazel/commit/b6f3111d8e460de35a5e5570bd784919fc30c5f8#diff-1f3936a41b59af9df49bf16acbe44524fbd503e8ed7cd76b62d5164867e0032aR160)? If that's the case, it's something I may figure out by myself and could submit as a follow-up.

Reply:
I had this in mind but didn't make any change to that line yet because I want to have the discussion in that PR and finalize how do we handle different commands. This is also the reason why this PR is still in draft.

Reply:
Now the test binaries are not downloaded with `bazel coverage`